# Tabular data with DataFrames.jl

The `result` that we got above is a `DataFrame` from [DataFrames.jl](http://juliadata.github.io/DataFrames.jl/stable/), one of the major Julia packages for tabular data. 


In [1]:
import Pkg
Pkg.activate(joinpath(@__DIR__, ".."))

  Activating project at `~/documents/presentations/IntroductionToJulia-API-2025`


In [2]:
using DataFrames, CSV

In this notebook, we'll be working on a small dataset from [Kaggle](https://www.kaggle.com/datasets/deepu1109/star-dataset?resource=download), which contains observed features of a few stars. The data is stored in a `CSV` file, which we'll first read into a `DataFrame` object.

In [5]:
filepath = joinpath(@__DIR__, "star_features.csv") # path of data file
df = CSV.read(filepath, DataFrame) # read the file and convert it to a `DataFrame` 

240×7 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  absolute_magnitude  s ⋯
     │ Int64               Float64          Float64      Float64             I ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068         0.0024         0.17                16.12     ⋯
   2 │               3042         0.0005         0.1542              16.6
   3 │               2600         0.0003         0.102               18.7
   4 │               2800         0.0002         0.16                16.65
   5 │               1939         0.000138       0.103               20.06     ⋯
   6 │               2840         0.00065        0.11                16.98
   7 │               2637         0.00073        0.127               17.22
   8 │               2600         0.0004         0.096               17.4
  ⋮  │         ⋮                  ⋮              ⋮               ⋮             ⋱
 234 │              27739    849420.0         1252.0                 -7.59     ⋯
 235 │              21904    748490.0         1130.0                 -7.67
 236 │              38940    374830.0         1356.0                 -9.93
 237 │              30839    834042.0         1194.0                -10.63
 238 │               8829    537493.0         1423.0                -10.73     ⋯
 239 │               9235    404940.0         1112.0                -11.23
 240 │              37882    294903.0         1783.0                 -7.8
                                                  3 columns and 225 rows omitted

We can see that this dataframe has 240 rows and 7 columns. To get the names of the columns, you can call the `names` function on the dataframe.

In [6]:
names(df)

7-element Vector{String}:
 "temperature_kelvin"
 "luminosity_Lsun"
 "radius_Rsun"
 "absolute_magnitude"
 "star_type"
 "star_colour"
 "spectral_class"

## Basic stuff
One can do basic selection of these data, for example selecting specific columns, removing columns, adding new columns or rows, etc. In this notebook we won't be using the  `star_type` column, so let's remove it:

In [7]:
select!(df, Not(:star_type))

240×6 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  absolute_magnitude  s ⋯
     │ Int64               Float64          Float64      Float64             S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068         0.0024         0.17                16.12   R ⋯
   2 │               3042         0.0005         0.1542              16.6    R
   3 │               2600         0.0003         0.102               18.7    R
   4 │               2800         0.0002         0.16                16.65   R
   5 │               1939         0.000138       0.103               20.06   R ⋯
   6 │               2840         0.00065        0.11                16.98   R
   7 │               2637         0.00073        0.127               17.22   R
   8 │               2600         0.0004         0.096               17.4    R
  ⋮  │         ⋮                  ⋮              ⋮               ⋮             ⋱
 234 │              27739    849420.0         1252.0                 -7.59   B ⋯
 235 │              21904    748490.0         1130.0                 -7.67   B
 236 │              38940    374830.0         1356.0                 -9.93   B
 237 │              30839    834042.0         1194.0                -10.63   B
 238 │               8829    537493.0         1423.0                -10.73   W ⋯
 239 │               9235    404940.0         1112.0                -11.23   W
 240 │              37882    294903.0         1783.0                 -7.8    B
                                                  2 columns and 225 rows omitted

To access specific rows or columns you can access a dataframe like a matrix, e.g. `df[:, 1:3]` gives the first three rows while `df[:, [:a, :b]]` gives the rows selected by name.

Adding new data to a dataframe is also straight forward. One can add new rows or new columns. New rows are added by either providing a vector (where the vector's length matches the number of columns), or by providing a named tuple that explicitly names which column gets which new value.

In [8]:
df1 = df[1:3, 1:3]

3×3 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun 
     │ Int64               Float64          Float64     
─────┼──────────────────────────────────────────────────
   1 │               3068           0.0024       0.17
   2 │               3042           0.0005       0.1542
   3 │               2600           0.0003       0.102

In [9]:
# add new row by order of columns
push!(df1, [5772, 1, 1])

4×3 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun 
     │ Int64               Float64          Float64     
─────┼──────────────────────────────────────────────────
   1 │               3068           0.0024       0.17
   2 │               3042           0.0005       0.1542
   3 │               2600           0.0003       0.102
   4 │               5772           1.0          1.0

In [10]:
# add new row by name of columnns
push!(df1, (temperature_kelvin = 3600, radius_Rsun=640, luminosity_Lsun = 65_000))

5×3 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun 
     │ Int64               Float64          Float64     
─────┼──────────────────────────────────────────────────
   1 │               3068           0.0024       0.17
   2 │               3042           0.0005       0.1542
   3 │               2600           0.0003       0.102
   4 │               5772           1.0          1.0
   5 │               3600       65000.0        640.0

In [12]:
# add new column with given name by field assignment
df1.random_columns = rand(size(df1, 1))
df1

5×5 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  z         random_colu ⋯
     │ Int64               Float64          Float64      Float64   Float64     ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068           0.0024       0.17    0.328993        0.310 ⋯
   2 │               3042           0.0005       0.1542  0.734847        0.990
   3 │               2600           0.0003       0.102   0.709464        0.405
   4 │               5772           1.0          1.0     0.583744        0.181
   5 │               3600       65000.0        640.0     0.425352        0.158 ⋯
                                                                1 column omitted

## Selecting specific rows

There are many ways to select specific rows based on values of certain columns. For example, if we wanted to select the stars with effective temperatures below a certain value, we could do:

In [14]:
df[df.temperature_kelvin .<= 5772, :] # select all the rows for which temperature_kelvin is less than or equal to 5772

120×6 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  absolute_magnitude  s ⋯
     │ Int64               Float64          Float64      Float64             S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068         0.0024         0.17                16.12   R ⋯
   2 │               3042         0.0005         0.1542              16.6    R
   3 │               2600         0.0003         0.102               18.7    R
   4 │               2800         0.0002         0.16                16.65   R
   5 │               1939         0.000138       0.103               20.06   R ⋯
   6 │               2840         0.00065        0.11                16.98   R
   7 │               2637         0.00073        0.127               17.22   R
   8 │               2600         0.0004         0.096               17.4    R
  ⋮  │         ⋮                  ⋮              ⋮               ⋮             ⋱
 114 │               3212         0.0016         0.378               12.854  R ⋯
 115 │               3523         0.0054         0.319               12.43   R
 116 │               3598         0.0027         0.67                13.667  R
 117 │               3142         0.00132        0.258               14.12   R
 118 │               3496         0.00125        0.336               14.94   R ⋯
 119 │               3324         0.0065         0.471               12.78   R
 120 │               3463         0.0027         0.675               14.776  R
                                                  2 columns and 105 rows omitted

In that example, we also included all the columns. If we wanted to only get specific columns, we could do

In [16]:
df[df.temperature_kelvin .<= 5772, [:luminosity_Lsun, :radius_Rsun]] # same as above, except we now only select two columns

120×2 DataFrame
 Row │ luminosity_Lsun  radius_Rsun 
     │ Float64          Float64     
─────┼──────────────────────────────
   1 │        0.0024         0.17
   2 │        0.0005         0.1542
   3 │        0.0003         0.102
   4 │        0.0002         0.16
   5 │        0.000138       0.103
   6 │        0.00065        0.11
   7 │        0.00073        0.127
   8 │        0.0004         0.096
  ⋮  │        ⋮              ⋮
 114 │        0.0016         0.378
 115 │        0.0054         0.319
 116 │        0.0027         0.67
 117 │        0.00132        0.258
 118 │        0.00125        0.336
 119 │        0.0065         0.471
 120 │        0.0027         0.675
                    105 rows omitted

We can also combine conditional statements to make more specific selections.

In [20]:
# select rows where temperature <= 5772, and radius is bigger than 0.11 and smaller than 0.67,
# and exclude the `star_colour` column
df[df.temperature_kelvin .<= 5772 .&& (0.11 .< df.radius_Rsun .< 0.67), Not(:star_colour)]

52×5 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  absolute_magnitude  s ⋯
     │ Int64               Float64          Float64      Float64             S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068         0.0024         0.17                16.12   M ⋯
   2 │               3042         0.0005         0.1542              16.6    M
   3 │               2800         0.0002         0.16                16.65   M
   4 │               2637         0.00073        0.127               17.22   M
   5 │               2700         0.00018        0.13                16.05   M ⋯
   6 │               3600         0.0029         0.51                10.69   M
   7 │               3129         0.0122         0.3761              11.79   M
   8 │               3134         0.0004         0.196               13.21   M
  ⋮  │         ⋮                  ⋮              ⋮               ⋮             ⋱
  46 │               3257         0.0024         0.46                10.73   M ⋯
  47 │               2994         0.0072         0.28                13.45   M
  48 │               3212         0.0016         0.378               12.854  M
  49 │               3523         0.0054         0.319               12.43   M
  50 │               3142         0.00132        0.258               14.12   M ⋯
  51 │               3496         0.00125        0.336               14.94   M
  52 │               3324         0.0065         0.471               12.78   M
                                                    1 column and 37 rows omitted

## Querying and manipulating a `DataFrame`

General querying and data manipulation can be done on `DataFrame`s using standard library functions from DataFrames.jl. However, it is [recommended](https://dataframes.juliadata.org/stable/man/querying_frameworks/#Data-manipulation-frameworks) to use a package dedicated to querying and data manupulation that provides simpler/easier syntax for it. Several options exist, each providing a slightly different variant on the syntax, so you can pick whatever you feel most comfortable in. Here we will use a LINQ-like [Query.jl](https://www.queryverse.org/Query.jl/stable/) syntax, which can be used with any tabular Julia data.

In [21]:
using Query
df

240×6 DataFrame
 Row │ temperature_kelvin  luminosity_Lsun  radius_Rsun  absolute_magnitude  s ⋯
     │ Int64               Float64          Float64      Float64             S ⋯
─────┼──────────────────────────────────────────────────────────────────────────
   1 │               3068         0.0024         0.17                16.12   R ⋯
   2 │               3042         0.0005         0.1542              16.6    R
   3 │               2600         0.0003         0.102               18.7    R
   4 │               2800         0.0002         0.16                16.65   R
   5 │               1939         0.000138       0.103               20.06   R ⋯
   6 │               2840         0.00065        0.11                16.98   R
   7 │               2637         0.00073        0.127               17.22   R
   8 │               2600         0.0004         0.096               17.4    R
  ⋮  │         ⋮                  ⋮              ⋮               ⋮             ⋱
 234 │              27739    849420.0         1252.0                 -7.59   B ⋯
 235 │              21904    748490.0         1130.0                 -7.67   B
 236 │              38940    374830.0         1356.0                 -9.93   B
 237 │              30839    834042.0         1194.0                -10.63   B
 238 │               8829    537493.0         1423.0                -10.73   W ⋯
 239 │               9235    404940.0         1112.0                -11.23   W
 240 │              37882    294903.0         1783.0                 -7.8    B
                                                  2 columns and 225 rows omitted

Let's perform a basic query that we explain step-by-step:

In [23]:
# `@from ... begin` initializes a query
# here `row` (any name would be fine) is the
# iterable of the tabular data. Think of it as a
# NamedTuple. You can access its values by name with the . syntax
q = @from row in df begin
    # `@where` filters elements where the following
    # expression is true.
    @where row.spectral_class == "O"
    # `@select {stuff... } creates a new named tuple
    @select {row.radius_Rsun, row.absolute_magnitude, Lsquared = row.luminosity_Lsun^2}
    # `@collect` collects the selected results into
    # the specified data structure
    @collect DataFrame
end

40×3 DataFrame
 Row │ radius_Rsun  absolute_magnitude  Lsquared   
     │ Float64      Float64             Float64    
─────┼─────────────────────────────────────────────
   1 │        10.6              -4.7    4.1616e10
   2 │        10.2              -4.4    3.9204e10
   3 │        14.0              -6.23   6.60969e11
   4 │        36.0              -5.76   1.6129e10
   5 │        83.0              -6.89   5.5225e10
   6 │        12.0              -5.91   1.2544e10
   7 │        42.0              -6.12   5.3361e10
   8 │        63.0              -5.83   1.31769e11
  ⋮  │      ⋮               ⋮               ⋮
  34 │        86.0              -5.905  1.20979e11
  35 │        92.0              -7.346  5.60836e11
  36 │      1356.0              -9.29   7.44362e10
  37 │      1948.5             -10.84   4.20461e11
  38 │      1356.0              -9.93   1.40498e11
  39 │      1194.0             -10.63   6.95626e11
  40 │      1783.0              -7.8    8.69678e10
                                    25 rows omitted

As you can see, you don't have to necessarily collect only existing columns, you can even create new ones (specified by name).

Of course, much, much more things are possible to do in such a query. For more, see the documentation of Query.jl.

# Exercises


## Plotting subsets of a dataframe

`DataFrames` has a default dataset that is used in its test suite. Install the `CSV` package, and load this dataset with the command:
```julia
using DataFrames, CSV
iris = DataFrame(CSV.File(
    joinpath(dirname(pathof(DataFrames)), 
    "../docs/src/assets/iris.csv")
))
```

This dataset has various flower species (column `:Species`). For every species create a 1x2 figure with the following plots:

* [1,1] = scatter plot of `SepalLength` vs `SepalWidth`. 
* [1,2] = scatter plot of `PetalLength` vs `PetalWidth`.

for each of these sub-scatter plots, calculate and print the Pearson correlaton coefficient.
